youtube_dl/extractor/viu.py

```diff
@@ -223,20 +269,64 @@ def _real_extract(self, url):
         if not video_data:
             raise ExtractorError('This video is not available in your region.', expected=True)
 
-        stream_data = self._download_json(
-            'https://d1k2us671qcoau.cloudfront.net/distribute_web_%s.php' % country_code,
-            video_id, 'Downloading stream info', query={
-                'ccs_product_id': video_data['ccs_product_id'],
-            }, headers={
-                'Referer': url,
-                'Origin': re.search(r'https?://[^/]+', url).group(0),
-            })['data']['stream']
+        duration_limit = False
+        query = {
+            'ccs_product_id': video_data['ccs_product_id'],
+            'language_flag_id': self._LANGUAGE_FLAG.get(lang_code.lower()) or '3',
+        }
+        headers = {
+            'Referer': re.search(r'https?://[^/]+', url).group(0),
+            'Origin': re.search(r'https?://[^/]+', url).group(0),
+        }
+        try:
+            stream_data = self._download_json(
+                'https://d1k2us671qcoau.cloudfront.net/distribute_web_%s.php' % country_code,
+                video_id, 'Downloading stream info', query=query, headers=headers)
+            stream_data = self._detect_error(stream_data)['stream']
+        except (ExtractorError, KeyError):
+            stream_data = None
+            if video_data.get('user_level', 0) > 0:
+                user = self._login(country_code, video_id)
+                if user:
+                    query['identity'] = user['identity']
+                    stream_data = self._download_json(
+                        'https://d1k2us671qcoau.cloudfront.net/distribute_web_%s.php' % country_code,
+                        video_id, 'Downloading stream info', query=query, headers=headers)
+                    stream_data = self._detect_error(stream_data).get('stream')
+                else:
+                    # preview is limited to 3min for non-members
+                    # try to bypass the duration limit
+                    duration_limit = True
+                    query['duration'] = '180'
+                    stream_data = self._download_json(
+                        'https://d1k2us671qcoau.cloudfront.net/distribute_web_%s.php' % country_code,
+                        video_id, 'Downloading stream info', query=query, headers=headers)
```

Comment:
You could avoid to use the exact same code twice here since all that changes seems to be the query. So you can alter the query in `if` statements and do the json download outside of it.

Reply:
actually the really repeating code is only one line, so here I extracted that specific line into a separated function:
https://github.com/blackjack4494/youtube-dlc/pull/211/commits/2290860e1f2d66f8c2ff373c1f7148c65c017077